audition/distance_from_best.py

```diff
@@ -0,0 +1,300 @@
+                        m.model_group_id,
+                        m.model_id,
+                        m.train_end_time,
+                        ev.value,
+                        row_number() OVER (
+                            PARTITION BY m.train_end_time
+                            ORDER BY ev.value DESC, RANDOM()
+                        ) AS rank
+                  FROM results.evaluations ev
+                  JOIN results.{models_table} m USING(model_id)
+                  JOIN results.model_groups mg USING(model_group_id)
+                  WHERE ev.metric='{metric}' AND ev.parameter='{metric_param}'
+                        AND m.model_group_id IN ({model_group_ids})
+                        AND train_end_time in ({train_end_times})
+                ),
+                model_tols AS (
+                  SELECT train_end_time, model_group_id, model_id,
+                         rank,
+                         value,
+                         first_value(value) over (
+                            partition by train_end_time
+                            order by rank ASC
+                        ) AS best_val
+                  FROM model_ranks
+                ),
+                current_best_vals as (
+                    SELECT
+                        model_group_id,
+                        model_id,
+                        train_end_time,
+                        '{metric}',
+                        '{metric_param}',
+                        value as raw_value,
+                        best_val - value below_best
+                    FROM model_tols
+                )
+                select
+                    current_best_vals.*,
+                    first_value(below_best) over (
+                        partition by model_group_id
+                        order by train_end_time asc
+                        rows between 1 following and unbounded following
+                    ) below_best_next_time
+                from current_best_vals
+            '''.format(
+                model_group_ids=str_in_sql(model_group_ids),
+                train_end_times=str_in_sql(train_end_times),
+                models_table=self.models_table,
+                metric=metric['metric'],
+                metric_param=metric['param'],
+                new_table=self.distance_table
+            ))
+
+    def create_and_populate(
+        self,
+        model_group_ids,
+        train_end_times,
+        metrics,
+    ):
+        """Creates and populates the distance table with the
+            given model groups, times, and metrics
+
+        Args:
+            model_group_ids (list) Model group ids to include in the distance table
+            train_end_times (list) Train end times to include in the table
+            metrics (list) Metrics and params to include in the table. Each
+                row should be a dict with keys:
+                        'metric' (e.g. 'precision@')
+                        'param' (e.g. '100_abs')
```

Comment:
I'm tempted to use 'parameter' to match the evaluations table, but it just seems...too generic?  What do you think?